service/history/timer_queue_active_task_executor.go

```diff
@@ -243,11 +243,8 @@ func (t *timerQueueActiveTaskExecutor) executeActivityTimeoutTask(
 	isHeartBeatTask := task.TimeoutType == enumspb.TIMEOUT_TYPE_HEARTBEAT
 	ai, heartbeatTimeoutVis, ok := mutableState.GetActivityInfoWithTimerHeartbeat(task.EventID)
 	if isHeartBeatTask && ok && queues.IsTimeExpired(task.GetVisibilityTime(), heartbeatTimeoutVis) {
-		err := mutableState.UpdateActivity(ai.ScheduledEventId, func(activityInfo *persistencespb.ActivityInfo, _ workflow.MutableState) error {
-			activityInfo.TimerTaskStatus = activityInfo.TimerTaskStatus &^ workflow.TimerTaskStatusCreatedHeartbeat
-			return nil
-		})
-		if err != nil {
+		ai.TimerTaskStatus = ai.TimerTaskStatus &^ workflow.TimerTaskStatusCreatedHeartbeat
```

Comment:
hmm I don't think we should change `ai` outside mutable state. Consider it's read only.